pyo3-derive-backend/src/utils.rs

```diff
@@ -20,9 +21,110 @@ pub fn if_type_is_python(ty: &syn::Type) -> bool {
     }
 }
 
+pub fn is_text_signature_attr(attr: &syn::Attribute) -> bool {
+    attr.path.is_ident("text_signature")
+}
+
+fn parse_text_signature_attr<T: Display + quote::ToTokens + ?Sized>(
+    attr: &syn::Attribute,
+    python_name: &T,
+    text_signature_line: &mut Option<syn::LitStr>,
+) -> syn::Result<Option<()>> {
+    if !is_text_signature_attr(attr) {
+        return Ok(None);
+    }
+    if text_signature_line.is_some() {
+        return Err(syn::Error::new_spanned(
+            attr,
+            "text_signature attribute already specified previously",
+        ));
+    }
+    let python_name_str = python_name.to_string();
+    let python_name_str = python_name_str
+        .rsplit('.')
+        .next()
+        .map(str::trim)
+        .filter(|v| !v.is_empty())
+        .ok_or_else(|| {
+            syn::Error::new_spanned(
+                &python_name,
+                format!("failed to parse python name: {}", python_name),
+            )
+        })?;
+    match attr.parse_meta()? {
+        syn::Meta::NameValue(syn::MetaNameValue {
+            lit: syn::Lit::Str(lit),
+            ..
+        }) => {
+            let value = lit.value();
+            if !value.starts_with('(') {
+                return Err(syn::Error::new_spanned(
+                    lit,
+                    "text_signature must start with \"(\"",
+                ));
+            }
+            if !value.ends_with(')') {
+                return Err(syn::Error::new_spanned(
+                    lit,
+                    "text_signature must end with \")\"",
+                ));
+            }
+            *text_signature_line = Some(syn::LitStr::new(
+                &(python_name_str.to_owned() + &value),
+                lit.span(),
+            ));
+        }
+        meta => {
+            return Err(syn::Error::new_spanned(
+                meta,
+                "text_signature must be of the form #[text_signature = \"\"]",
+            ));
+        }
+    };
+    Ok(Some(()))
+}
+
+pub fn parse_text_signature_attrs<T: Display + quote::ToTokens + ?Sized>(
+    attrs: &mut Vec<syn::Attribute>,
```

Comment:
You really need `&mut` and `retain` for this function?

Reply:
fixed